README.md
# Reliablereq
[![Build Status](https://travis-ci.org/globocom/reliable-request.svg?branch=master)](https://travis-ci.org/globocom/reliable-request) [![Go Report Card](https://goreportcard.com/badge/github.com/globocom/reliable-request)](https://goreportcard.com/report/github.com/globocom/reliable-request)

A golang opinionated library to provide reliable request using [hystrix-go](https://github.com/afex/hystrix-go), [go-cache](https://github.com/patrickmn/go-cache), and [go-resiliency](https://github.com/eapache/go-resiliency).

When you do a `Get`, it provides:
* an HTTP client configured with timeouts and [keepalive](https://en.wikipedia.org/wiki/HTTP_persistent_connection),
* a [circuit breaker](https://martinfowler.com/bliki/CircuitBreaker.html) and
* a proper [caching system](https://en.wikipedia.org/wiki/Cache_(computing)).

# Usage

```golang
req := reliablereq.NewReliableRequest()
req.TTLCache = 1 * time.Second
req.EnableStaleCache = false
body, err := req.Get("http://example.com/list")

// passing authentication/authorization bearer token
req := reliablereq.NewReliableRequest()
req.Headers = map[string]string{"Authorization": "Bearer foobar"}
body, err := req.Get("http://example.com/list")

// creating a different hystrix command
req := reliablereq.NewReliableRequest()
req.UpdateHystrixConfig("api2command", hystrix.CommandConfig{
		Timeout:                800 + 100,
		MaxConcurrentRequests:  100,
		ErrorPercentThreshold:  50,
		RequestVolumeThreshold: 20,
		SleepWindow:            5000,
	})
 body, err := req.Get("http://example.com/list")
```

## WARNING
Make sure you use different Hystrix commands for other endpoint APIs or separated Circuit Breaker contexts, otherwise, an endpoint may open the circuit breaker and all other requests will fail.


# Opinionated defaults

```golang
// reliable request defaults
rr := ReliableRequest{
  EnableCache:        true,
  TTLCache:           1 * time.Minute,
  EnableStaleCache:   true,
  TTLStaleCache:      24 * time.Hour,
}
// hystrix
var defaultHystrixConfiguration = hystrix.CommandConfig{
  Timeout:                800 + 100, // the defaultTimeout http client + a small gap
  MaxConcurrentRequests:  100,
  ErrorPercentThreshold:  50,
  RequestVolumeThreshold: 3,
  SleepWindow:            5000,
}
// http client
client := &http.Client{
  Transport: &http.Transport{
    DialContext: (&net.Dialer{
      Timeout:   800 * time.Millisecond,
      KeepAlive: 30 * time.Second,
    }).DialContext,
    MaxIdleConns:        100,
    MaxIdleConnsPerHost: 100,
    TLSHandshakeTimeout: 800 * time.Millisecond,
  },
  Timeout: 800 * time.Millisecond,
}
```

# Future

* provide a proxy to setup hystrix
* add retry logic (by go-resiliency)
* add more examples, like token header requests and more
* discuss the adopted defaults
* discuss whether async hystrix is better (Go instead of Do)
* understand and test the simultaneous client req hystrix config to see its implications
* add go api documentation
* add hooks (callbacks) to provides means for metrics gathering
* add more HTTP verbs?
* add load stress


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.2] - 2020-08-27
### Changed
- Expose cache data

## [0.0.1] - 2019-07-16
### Added
- Use histryx-go as circuit breaker
- Add cache feature
- Add stale feature

[Unreleased]: https://github.com/globocom/reliable-request/compare/v0.0.1...HEAD
[0.0.1]: https://github.com/globocom/reliable-request/releases/tag/v0.0.1


reliablereq/reliablereq_test.go
package reliablereq

import (
	"testing"
	"time"

	"github.com/afex/hystrix-go/hystrix"
	"github.com/stretchr/testify/assert"
	gock "gopkg.in/h2non/gock.v1"
)

func Test_It_returns_a_valid_response(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	// we need to intercept current http client due
	// https://github.com/h2non/gock/issues/27#issuecomment-334177773
	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)
}

func Test_It_raises_an_error_when_there_is_no_connection(t *testing.T) {
	Flush()

	req := NewReliableRequest()

	_, err := req.Get("http://example.non/list")

	assert.NotNil(t, err)
}

func Test_It_returns_an_error_when_server_responds_with_a_non_2xx(t *testing.T) {

	defer gock.Off()
	Flush()

	non2xx := []int{400, 404, 503}
	for _, status := range non2xx {
		gock.New("http://example.com").
			Get("/list").
			Reply(status)

		req := NewReliableRequest()
		gock.InterceptClient(req.HTTPClient)
		defer gock.RestoreClient(req.HTTPClient)

		_, err := req.Get("http://example.com/list")

		assert.NotNil(t, err)
	}
}

func Test_It_uses_cache_when_enabled(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Times(1).
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)

	body, err = req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)
}

func Test_It_doesnt_use_cache_when_disabled(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Times(1).
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	req.EnableCache = false
	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)

	body, err = req.Get("http://example.com/list")

	assert.NotNil(t, err)
}

func Test_It_uses_stale_cache_when_enabled(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Times(1).
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	req.TTLCache = 1 * time.Second
	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)

	// simulating a cache eviction
	time.Sleep(2 * time.Second)

	body, err = req.Get("http://example.com/list")

	assert.Nil(t, err)
}

func Test_It_doesnt_use_stale_cache_when_disabled(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Times(1).
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	req.TTLCache = 1 * time.Second
	req.EnableStaleCache = false
	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)

	// simulating a cache eviction
	time.Sleep(2 * time.Second)

	body, err = req.Get("http://example.com/list")

	assert.NotNil(t, err)
}

func Test_It_allows_custom_headers(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		MatchHeader("Authorization", "^foo bar$").
		Get("/list").
		Times(1).
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	req.Headers = map[string]string{"Authorization": "foo bar"}

	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, err := req.Get("http://example.com/list")

	assert.Nil(t, err)
	assert.NotNil(t, body)
	assert.Equal(t, "{\"name\":\"mock\"}\n", body)
}
func Test_It_opens_the_circuit_breaker_when_error_percentage_is_reached(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list").
		Times(5).
		Reply(503)

	req := NewReliableRequest()
	req.UpdateHystrixConfig("custom_cb", hystrix.CommandConfig{
		Timeout:                800 + 100, // the defaultTimeout http client + a small gap
		MaxConcurrentRequests:  100,
		ErrorPercentThreshold:  50,
		RequestVolumeThreshold: 4,
		SleepWindow:            5000,
	})

	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	body, _ := req.Get("http://example.com/list")
	body, _ = req.Get("http://example.com/list")
	body, _ = req.Get("http://example.com/list")
	body, _ = req.Get("http://example.com/list")
	body, _ = req.Get("http://example.com/list")

	cb, _, _ := hystrix.GetCircuit("custom_cb")

	assert.Equal(t, "", body)
	assert.Equal(t, true, cb.IsOpen())
}
func Test_It_closes_the_circuit_breaker_after_the_sleep_window(t *testing.T) {
	defer gock.Off()
	Flush()

	gock.New("http://example.com").
		Get("/list0").
		Reply(503)

	gock.New("http://example.com").
		Get("/list1").
		Reply(503)

	gock.New("http://example.com").
		Get("/list2").
		Reply(503)

	gock.New("http://example.com").
		Get("/list3").
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	gock.New("http://example.com").
		Get("/list4").
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	gock.New("http://example.com").
		Get("/list5").
		Reply(200).
		JSON(map[string]interface{}{"name": "mock"})

	req := NewReliableRequest()
	req.UpdateHystrixConfig("custom_cb", hystrix.CommandConfig{
		Timeout:                800 + 100, // the defaultTimeout http client + a small gap
		MaxConcurrentRequests:  100,
		ErrorPercentThreshold:  50,
		RequestVolumeThreshold: 2,
		SleepWindow:            2000,
	})

	gock.InterceptClient(req.HTTPClient)
	defer gock.RestoreClient(req.HTTPClient)

	cb, _, _ := hystrix.GetCircuit("custom_cb")

	body, _ := req.Get("http://example.com/list0") //error 100%
	body, _ = req.Get("http://example.com/list1")  //error 100%

	assert.Equal(t, "", body)
	assert.Equal(t, true, cb.IsOpen()) // open due to error percentage + request volume threshold

	body, _ = req.Get("http://example.com/list2") //error 100%
	body, _ = req.Get("http://example.com/list3") //error 75%

	time.Sleep(3500 * time.Millisecond) // simulating its sleep window

	body, _ = req.Get("http://example.com/list4") //error 60%
	body, _ = req.Get("http://example.com/list5") //error 50%

	assert.Equal(t, false, cb.IsOpen()) // closed due to error percentage
}


reliable_request.go
package reliablereq

import (
	"fmt"
	"net/http"
	"time"

	"github.com/afex/hystrix-go/hystrix"
	"github.com/imdario/mergo"
	"github.com/patrickmn/go-cache"
	"github.com/pkg/errors"
)

var cachedResponses *cache.Cache

func init() {
	makeDefaultHistryx()
	cachedResponses = cache.New(5*time.Minute, 10*time.Minute)
}

// ReliableRequest - a struct holding params to make reliable requests
type ReliableRequest struct {
	Headers            map[string]string
	HTTPClient         *http.Client
	EnableCache        bool
	TTLCache           time.Duration
	EnableStaleCache   bool
	TTLStaleCache      time.Duration
	HystrixCommandName string
	HystrixCommand     hystrix.CommandConfig
}

// NewReliableRequest - create a new ReliableRequest
func NewReliableRequest(rr ReliableRequest) *ReliableRequest {
	mergo.Merge(&rr, defaultReliableRequest())

	if rr.HystrixCommandName != defaultCommandName {
		hystrix.ConfigureCommand(rr.HystrixCommandName, rr.HystrixCommand)
	}

	return &rr
}

// Request - method to request data from a url
func (rr *ReliableRequest) Request(url string) (*http.Response, error) {
	var resp *http.Response

	if rr.EnableCache {
		cached, found := rr.getCache(url)
		if found {
			return cached, nil
		}
	}

	err := hystrix.Do(rr.HystrixCommandName, func() error {
		var err error
		resp, err = rr.urlRequest(url)
		if err == nil {
			rr.setCache(url, resp)
		}
		return err
	}, func(previousError error) error {
		var err error

		if rr.EnableStaleCache {
			resp, err = rr.getStaleCache(url)
			if err != nil {
				return errors.Wrap(previousError, err.Error())
			}
		} else {
			return previousError
		}

		return nil
	})

	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("could not complete the request for url"))
	}

	return resp, nil
}

func keyStale(key string) string {
	return fmt.Sprintf("%s-stale", key)
}

func (rr *ReliableRequest) setCache(url string, resp *http.Response) {
	cachedResponses.Set(url, resp, rr.TTLCache)

	if rr.EnableStaleCache {
		cachedResponses.Set(keyStale(url), resp, rr.TTLStaleCache)
	}
}

func (rr *ReliableRequest) getCache(key string) (*http.Response, bool) {
	cached, found := cachedResponses.Get(key)
	if found {
		return cached.(*http.Response), true
	}

	return nil, false
}

func (rr *ReliableRequest) getStaleCache(key string) (*http.Response, error) {
	cached, found := rr.getCache(keyStale(key))
	if found {
		return cached, nil
	}
	return nil, errors.New("failed to fetch stale response")
}

func (rr *ReliableRequest) urlRequest(url string) (*http.Response, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("could not create an http request for url: %s", url))
	}

	if rr.Headers != nil {
		for k, v := range rr.Headers {
			req.Header.Add(k, v)
		}
	}
	resp, err := rr.HTTPClient.Do(req)

	if err == nil {
		// we only cache 2xx
		if resp.StatusCode != http.StatusOK {
			return nil, errors.New(fmt.Sprintf("bad response: %s for url: %s", resp.Status, url))
		}
	}

	return resp, err
}


reliablereq/reliablereq.go
package reliablereq

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"time"

	"github.com/afex/hystrix-go/hystrix"
	"github.com/patrickmn/go-cache"
	"github.com/pkg/errors"
)

var cachedResponses *cache.Cache

func init() {
	makeDefaultHistryx()
	cachedResponses = cache.New(5*time.Minute, 10*time.Minute)
}

func Flush() {
	cachedResponses.Flush()
	hystrix.Flush()
}

// ReliableRequest - a struct holding params to make reliable requests
type ReliableRequest struct {
	Headers            map[string]string
	HTTPClient         *http.Client
	EnableCache        bool
	TTLCache           time.Duration
	EnableStaleCache   bool
	TTLStaleCache      time.Duration
	hystrixCommandName string
}

// NewReliableRequest - create a new ReliableRequest
func NewReliableRequest() *ReliableRequest {
	rr := defaultReliableRequest()

	return &rr
}

// Get - returns the requested data as string and a possible error
func (rr *ReliableRequest) Get(url string) (string, error) {
	var body string

	if rr.EnableCache {
		cached, found := rr.GetCache(url)
		if found {
			return cached, nil
		}
	}

	err := hystrix.Do(rr.hystrixCommandName, func() error {
		resp, err := rr.urlRequest(url)
		if err == nil {
			rawBody, _ := ioutil.ReadAll(resp.Body)
			body = string(rawBody)
			rr.setCache(url, body)
		}
		return err
	}, func(previousError error) error {
		var err error

		if rr.EnableStaleCache {
			body, err = rr.getStaleCache(url)
			if err != nil {
				return errors.Wrap(previousError, err.Error())
			}
		} else {
			return previousError
		}

		return nil
	})

	if err != nil {
		return "", errors.Wrap(err, fmt.Sprintf("could not complete the request for url"))
	}
	return body, nil
}

//UpdateHystrixConfig - configure a new circuit breaker
func (rr *ReliableRequest) UpdateHystrixConfig(name string, conf hystrix.CommandConfig) {
	rr.hystrixCommandName = name
	hystrix.ConfigureCommand(name, conf)
}

func keyStale(key string) string {
	return fmt.Sprintf("%s-stale", key)
}

func (rr *ReliableRequest) setCache(url, body string) {
	cachedResponses.Set(url, body, rr.TTLCache)

	if rr.EnableCache && rr.EnableStaleCache {
		cachedResponses.Set(keyStale(url), body, rr.TTLStaleCache)
	}
}

func (rr *ReliableRequest) GetCache(key string) (string, bool) {
	cached, found := cachedResponses.Get(key)
	if found {
		return cached.(string), true
	}

	return "", false
}

func (rr *ReliableRequest) getStaleCache(key string) (string, error) {
	cached, found := rr.GetCache(keyStale(key))
	if found {
		return cached, nil
	}
	return "", errors.New("failed to fetch stale response")
}

func (rr *ReliableRequest) urlRequest(url string) (*http.Response, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, errors.Wrap(err, fmt.Sprintf("could not create an http request for url: %s", url))
	}

	if rr.Headers != nil {
		for k, v := range rr.Headers {
			req.Header.Add(k, v)
		}
	}
	resp, err := rr.HTTPClient.Do(req)

	if err == nil {
		// we only cache 2xx
		if resp.StatusCode != http.StatusOK {
			return nil, errors.New(fmt.Sprintf("bad response: %s for url: %s", resp.Status, url))
		}
	}

	return resp, err
}


reliablereq/default.go
package reliablereq

import (
	"net"
	"net/http"
	"time"

	"github.com/afex/hystrix-go/hystrix"
)

const defaultTCPConnectionTimeout = 500 * time.Millisecond
const defaultTimeout = 800 * time.Millisecond
const defaultKeepAliveConnections = 100
const defaultCommandName = "default_config"

func timeoutHTTPClient() *http.Client {
	client := &http.Client{
		Transport: &http.Transport{
			DialContext: (&net.Dialer{
				Timeout:   defaultTCPConnectionTimeout,
				KeepAlive: 30 * time.Second,
			}).DialContext,
			MaxIdleConns:        defaultKeepAliveConnections,
			MaxIdleConnsPerHost: defaultKeepAliveConnections,
			TLSHandshakeTimeout: defaultTCPConnectionTimeout,
		},
		Timeout: defaultTimeout,
	}
	return client
}

func makeDefaultHistryx() {
	var defaultHystrixConfiguration = hystrix.CommandConfig{
		Timeout:                800 + 100, // the defaultTimeout http client + a small gap
		MaxConcurrentRequests:  100,
		ErrorPercentThreshold:  50,
		RequestVolumeThreshold: 20,
		SleepWindow:            5000,
	}

	hystrix.ConfigureCommand(defaultCommandName, defaultHystrixConfiguration)
}

func defaultReliableRequest() ReliableRequest {
	rr := ReliableRequest{
		HTTPClient:         timeoutHTTPClient(),
		EnableCache:        true,
		TTLCache:           1 * time.Minute,
		EnableStaleCache:   true,
		TTLStaleCache:      24 * time.Hour,
		hystrixCommandName: defaultCommandName,
	}
	return rr
}
